nixos/modules/services/web-servers/traefik.nix

```diff
@@ -4,56 +4,111 @@ with lib;
 
 let
   cfg = config.services.traefik;
-  configFile =
-    if cfg.configFile == null then
+  jsonValue = with types; let
+    valueType = nullOr (oneOf [
+      bool
+      int
+      float
+      str
+      (lazyAttrsOf valueType)
+      (listOf valueType)
+    ]) // {
+      description = "JSON value";
+      emptyValue.value = {};
+    };
+  in valueType;
+  dynamicConfigFile =
+    if cfg.dynamicConfigFile == null then
       pkgs.runCommand "config.toml" {
         buildInputs = [ pkgs.remarshal ];
         preferLocalBuild = true;
       } ''
         remarshal -if json -of toml \
-          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
+          < ${pkgs.writeText "dynamic_config.json" (builtins.toJSON cfg.dynamicConfigOptions)} \
+          > $out
+      ''
+    else cfg.dynamicConfigFile;
+  recursiveMerge = attrList:
+    let f = attrPath:
+      zipAttrsWith (n: values:
+        if tail values == []
+        then head values
+        else
+          if all isList values
+          then unique (concatLists values)
+          else
+            if all isAttrs values
+            then f (attrPath ++ [n]) values
+            else last values
+      );
+    in f [] attrList;
+  staticConfigFile =
+    if cfg.staticConfigFile == null then
+      pkgs.runCommand "config.toml" {
+        buildInputs = [ pkgs.yj ];
+        preferLocalBuild = true;
+      } ''
+        yj -jt -i \
+          < ${pkgs.writeText "static_config.json" (builtins.toJSON (recursiveMerge [cfg.staticConfigOptions { providers.file.filename = "${dynamicConfigFile}";}]))} \
```

Comment:
Can you break this line? It's really looong.

Reply:
Agreed, I've used nixfmt to format the whole file.